src/validation.cpp

```diff
@@ -1024,46 +993,104 @@ bool MemPoolAccept::Finalize(ATMPArgs& args, Workspace& ws)
+    MempoolAcceptResult result(ptx);
 
-    if (!PreChecks(args, workspace)) return false;
+    if (!PreChecks(args, result, workspace)) {
+        result.m_accepted = false;
+        return result;
+    }
 
     // Only compute the precomputed transaction data if we need to verify
     // scripts (ie, other policy checks pass). We perform the inexpensive
     // checks first and avoid hashing and signature verification unless those
     // checks pass, to mitigate CPU exhaustion denial-of-service attacks.
     PrecomputedTransactionData txdata;
 
-    if (!PolicyScriptChecks(args, workspace, txdata)) return false;
+    if (!PolicyScriptChecks(args, result, workspace, txdata)) {
+        result.m_accepted = false;
+        return result;
+    }
 
-    if (!ConsensusScriptChecks(args, workspace, txdata)) return false;
+    if (!ConsensusScriptChecks(args, result, workspace, txdata)) {
+        result.m_accepted = false;
+        return result;
+    }
 
     // Tx was accepted, but not added
-    if (args.m_test_accept) return true;
+    if (args.m_test_accept) {
+        result.m_accepted = true;
+        return result;
+    }
 
-    if (!Finalize(args, workspace)) return false;
+    if (!Finalize(args, result, workspace)) {
+        result.m_accepted = false;
+        return result;
+    }
 
     GetMainSignals().TransactionAddedToMempool(ptx, m_pool.GetAndIncrementSequence());
 
-    return true;
+    result.m_accepted = true;
+    return result;
+}
+
+std::vector<MempoolAcceptResult> MemPoolAccept::AcceptMultipleTransactions(std::vector<const CTransactionRef>& txns, ATMPArgs& args)
+{
+    AssertLockHeld(cs_main);
+    LOCK(m_pool.cs);
+    std::vector<MempoolAcceptResult> results{};
+    std::vector<Workspace> workspaces{};
+    workspaces.reserve(txns.size());
+    std::transform(txns.begin(), txns.end(), std::back_inserter(workspaces), [](const CTransactionRef& tx) {
+        return Workspace(tx);
+    });
+    // Do all PreChecks first and fail fast to avoid running expensive script
+    // checks when unnecessary.
+    for (unsigned int i = 0; i < txns.size(); ++i) {
+        MempoolAcceptResult result(txns[i]);
+        if (!PreChecks(args, result, workspaces[i])) {
+            result.m_accepted = false;
+            return std::vector<MempoolAcceptResult> {std::move(result)};
+        }
+        result.m_accepted = true; // Allowed so far
+        m_viewmempool.AddPackageTransaction(txns[i]);
+        results.push_back(result);
+    }
+
+    // TODO: Enforce package-level feerate and other policies before script checks.
+
+    for (unsigned int i = 0; i < txns.size(); ++i) {
+        PrecomputedTransactionData txdata;
+
+        if (!PolicyScriptChecks(args, results[i], workspaces[i], txdata)) {
+            results[i].m_accepted = false;
+            return std::vector<MempoolAcceptResult> {std::move(results[i])};
+        }
+
+        if (!ConsensusScriptChecks(args, results[i], workspaces[i], txdata)) {
```

Comment:
This means that we will update our script cache with transactions that might not ultimately get accepted to the mempool?  This would need to be restructured before exposing to the p2p network to avoid becoming a DoS vector.

Reply:
On further thought: if you're only implementing this for testmempoolaccept anyway, why bother with the call to ConsensusScriptChecks?  It can only return failure if our software is broken -- it's a safeguard against miners creating invalid blocks, not something that users should expect to ever run into.

If you drop this call, then dropping CIFMAC is also no longer necessary, right?

Reply:
AFAIK we only use the script cache when it passes, so I don't think it's a DoS vector? On master, if you call `testmempoolaccept`, it caches the successful script executions with consensus flags.

> On further thought: if you're only implementing this for testmempoolaccept anyway, why bother with the call to ConsensusScriptChecks?

Good point, `PolicyScriptChecks` is stricter so `ConsensusScriptChecks` is unnecessary. And yes, we could then keep CIFMAC. But in order to do actual, non-test-accepts, we'll have to revisit the issue again (since I assume we'll want to call `ConsensusScriptChecks` then). It'd be "kicking the can down the road."

Reply:
I assume that the ultimate goal of package validation is to only accept the whole set of transactions if each one individually makes it into the mempool (probably along with some other properties on the feerate of the package being used rather than individual feerates).  If that's correct, then calling `ConsensusScriptChecks` on anything (which will add to the script cache) before you finish calling `PolicyScriptChecks` on later transactions in the package will mean that a policy failure in a later transaction could cause the whole package to be rejected, while still having updated the cache.

If exposed on the p2p network, an attacker could wipe out a target's script cache for free this way, which is the DoS vector I was referring to.

So instead, you can make this loop just call `PolicyScriptChecks`, which is good enough for test_accept.  Once you implement adding to the mempool on success, you could have a new loop over the set of transactions in the package (topologically sorted) that invokes `ConsensusScriptChecks` and then adds the transaction to the mempool.  This way, CIFMAC doesn't need to be touched at all, because all a transaction's inputs are already in the utxo set or the mempool at the time `ConsensusScriptChecks` is invoked.  (This is essentially the structure I proposed in #16401.)

Reply:
> I assume that the ultimate goal of package validation is to only accept the whole set of transactions if each one individually makes it into the mempool... 
If that's correct, then calling ConsensusScriptChecks on anything (which will add to the script cache) before you finish calling PolicyScriptChecks on later transactions in the package will mean that a policy failure in a later transaction could cause the whole package to be rejected, while still having updated the cache.

Right, I have the same vision in mind, and absolutely agree. I should change these two be separate loops - all `PolicyScriptChecks`, then all `ConsensusScriptChecks`. This would mean that we would only be calling `ConsensusScriptChecks` (which is effectively just used to cache script results) if all of them would pass - there's no way for any of them to fail consensus if they passed policy.

Reply:
The more I think about this, the more it makes sense 😅 
In actual package accept, we don't need CoinsViewMempool after `PolicyScriptChecks`, we can actually just go one by one in the package (after topological sort) and run `ConsensusScriptChecks` + submit to mempool. And subsequent txns should have all of the Coins they need from the pool. Apologies if this was obvious to you - is this what you had in mind?

Reply:
Yes -- sounds like we're on the same page!